Add pull mode to remote job execution (#1630)
theforeman/foreman-documentation: guides/common/modules/proc_configuring-remote-execution-for-pull-client.adoc

= Configuring Remote Execution for Pull Client

By default, Remote Execution uses SSH as the transport mechanism for the Script provider.
However, Remote Execution also offers pull-based transport, which you can use if your infrastructure prohibits outgoing connections from {SmartProxy} to hosts.

> {Project} features `pull-mqtt` mode on {SmartProxy} with the pull client on hosts for pull-based transport.

Reviewer: ```suggestion
This is comprised of a `pull-mqtt` mode on {SmartProxy} in combination with a pull client running on hosts.
```

NOTE: The `pull-mqtt` mode works only with the Script provider.
Ansible and other providers will continue to use their default transport settings.

The mode is configured per {SmartProxy}.
Some {SmartProxies} can be configured to use `pull-mqtt` mode while others use SSH.
If this is the case, it is possible that one remote job on a given host will use the pull client and the next job on the same host will use SSH.
If you wish to avoid this scenario, configure all {SmartProxies} to use the same mode.

.Procedure
. Enable the pull-based transport on each relevant {SmartProxyServer}:
+
[options="nowrap" subs="quotes,attributes"]
----
 # {installer-scenario-smartproxy} \
--foreman-proxy-plugin-remote-execution-script-mode pull-mqtt
----
. Configure the firewall to allow MQTT service on port 1883:
+
[options="nowrap", subs="+quotes,verbatim,attributes"]
----
# firewall-cmd --add-port="1883/tcp"
# firewall-cmd --runtime-to-permanent
----
. In `pull-mqtt` mode, hosts subscribe for job notifications to the {SmartProxy} through which they are registered.
Therefore, it is recommended to ensure that {ProjectServer} sends remote execution jobs to that same {SmartProxy}.

> To do this, in the {ProjectWebUI} navigate to *Administer* > *Settings*.

Reviewer 2: ```suggestion
To do this, in the {ProjectWebUI}, navigate to *Administer* > *Settings*.
```

On the *Content* tab, set the value of *Prefer registered through {SmartProxy} for remote execution* to *Yes*.

> . After the pull-based transport is set up on a {SmartProxy}, it must also be configured on each host.

Reviewer 2: Please reword to make it active, aka. "you have to configure ...".

> For more information, see {ManagingHostsDocURL}transport-modes-for-remote-execution_managing-hosts[Transport Modes for Remote Execution] in _{ManagingHostsDocTitle}_.

Reviewer: Are we sure this shouldn't link directly to "Configuring a Host to Use the Pull Client"?

Author: Yes, because "Configuring" refers only to an existing host whereas "Transport Modes" lists several options for new hosts as well (in Additional resources). Additionally, "Configuring" is not published for non-Katello products, but the user can still use pull-mqtt even when they're not using Katello if I understand the matter correctly.

Reviewer: Yes that's true.  Okay, makes sense :+1:
